src/Bitcoin/Address/Bech32.hs

```diff
@@ -174,12 +180,36 @@ maxBech32Length = 90
 -- than 90 characters.
 bech32Encode :: Bech32Encoding -> HRP -> [Word5] -> Maybe Bech32
 bech32Encode enc hrp dat = do
-    guard $ checkHRP hrp
+    Bech32EncodeResult
+        { encodeResult
+        , encodeValidHrp = True
+        , encodeValidLength = True
+        } <-
+        pure $ bech32EncodeResult enc hrp dat
+    return encodeResult
+
+
+-- | The result of encoding a 'Bech32' string
+data Bech32EncodeResult = Bech32EncodeResult
+    { encodeResult :: Text
```

Comment:
Do we wanna make this `!Text`?

Reply:
I think there is a good case to be made for making this field lazy. After all consumers may check one of the error fields and decide to throw away the encoding.  Maybe we can reduce indirection for `bech32Encode` by changing the last line to `encodeResult `seq` pure encodeResult`?

Reply:
Is the reason for wanting `seq` there to reduce some unwanted laziness? I think the following would no longer terminate if `encodeResult` was somehow not total, where it would have previously otherwise, ie: 
```haskell
case bech32Encode input of { Just _ -> "Valid bech32"'; Nothing -> "Invalid bech32" }
```
Happy to go with that if we're all convinced that `encodeResult` is always total. I'm also fine with keeping existing definitions for bech32Encode / bech32Decode if we're worried about the extra overhead caused by the lazy data types.

Reply:
That was a misguided suggestion.  Let's leave things as is.  FWIW that making `encodeResult` a strict field has the same effect with respect to making `bech32Encode` diverge.

Reply:
My suggestion was rooted in the perhaps false assumption that we could predict that the overwhelming majority of the time we would want to read the decoded result out. If that isn't the case then we can leave it as is.